src/test.js

```diff
@@ -810,22 +811,106 @@ export function resetTestTimeout( timeoutDuration ) {
+//   uncaught error means we will no longer wait for the release (as it might not arrive).
+//
+// * Pause is never released, or called an insufficient number of times.
+//
+//   Our timeout handler will kill the pause and resume test processing, basically
+//   like internalRecover(), but for one pause instead of any/all.
+//
+//   Here, too, any late calls to resume() will be silently ignored to avoid
+//   extra errors. We tolerate this since the original test will have already been
+//   marked as failure.
+//
+//   TODO: QUnit 3 will enable timeouts by default <https://github.com/qunitjs/qunit/issues/1483>,
+//   but right now a test will hang indefinitely if async pauses are not released,
+//   unless QUnit.config.testTimeout or assert.timeout() is used.
+//
+// * Pause is spontaneously released during a different test,
+//   or when no test is currently running.
+//
+//   This is close to impossible because this error only happens if the original test
+//   succesfully finished first (since other failure scenarios kill pauses and ignore
+//   late calls). It can happen if a test ended exactly as expected, but has some
+//   external or shared state continuing to hold a reference to the release callback,
+//   and either the same test scheduled another call to it in the future, or a later test
+//   causes it to be called through some shared state.
+//
+// * Pause release() is called too often, during the same test.
+//
+//   This simply throws an error, after which uncaught error handling picks it up
+//   and processing resumes.
+export function internalStop( test, requiredCalls = 1 ) {
 	config.blocking = true;
 
-	// Set a recovery timeout, if so configured.
+	const pauseId = `#${test.asyncNextPauseId++}`;
+	const pause = {
+		killed: false,
+		remaining: requiredCalls
+	};
+	test.asyncPauses.set( pauseId, pause );
+
+	function release() {
+		if ( pause.killed ) {
+			return;
+		}
+		if ( config.current === undefined ) {
+			throw new Error( "Unexpected release of async pause after tests finished.\n" +
+				`> Test: ${test.testName} [async ${pauseId}]` );
+		}
+		if ( config.current !== test ) {
+			throw new Error( "Unexpected release of async pause during a different test.\n" +
+				`> Test: ${test.testName} [async ${pauseId}]` );
+		}
+		if ( pause.remaining <= 0 ) {
+			throw new Error( "Tried to release async pause that was already released.\n" +
+				`> Test: ${test.testName} [async ${pauseId}]` );
+		}
+
+		// The `requiredCalls` parameter exists to support `assert.async(count)`
+		pause.remaining--;
+		if ( pause.remaining === 0 ) {
+			test.asyncPauses.delete( pauseId );
+		}
+
+		internalStart( test );
+	}
+
 	if ( setTimeout ) {
 		let timeoutDuration;
-
 		if ( typeof test.timeout === "number" ) {
 			timeoutDuration = test.timeout;
 		} else if ( typeof config.testTimeout === "number" ) {
 			timeoutDuration = config.testTimeout;
 		}
 
+		// Set a recovery timeout, if so configured.
```

Comment:
```suggestion
	// Set a recovery timeout, if so configured.
	if ( setTimeout ) {
		let timeoutDuration;
		if ( typeof test.timeout === "number" ) {
			timeoutDuration = test.timeout;
		} else if ( typeof config.testTimeout === "number" ) {
			timeoutDuration = config.testTimeout;
		}

```

Reply:
I didn't quite get what you meant here. I had moved the inline comment to be closer to where "if configure" happens. Are you suggesting to leave it as-is? Or is there another difference?

Reply:
I hadn't noticed it matched the preexisting pattern because of the new code, but yes. Having the inline comment just before the check for `setTimeout` inherently explains why that check is there.

Reply:
I thought of the comment as explaining that `config.timeoutHandler` happens conditionally based on whether a timeout is configured (via one of the two means). The env check for setTimeout seems unrelated, but I don't disagree that reading it up front before mentally entering the overall block is also useful. If/when the block ends up doing multiple things, then I might push to move it back, but anyway, we're not there now. I'll leave it where it was.